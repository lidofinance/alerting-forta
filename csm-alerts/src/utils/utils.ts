import { Contract, EventFilter, Event } from 'ethers'

const LOG_FILTER_CHUNK = 2000

export async function getLogsByChunks(contract: Contract, filter: EventFilter, startblock: number, endBlock: number) {
  const events: Event[] = []
  let endBlockChunk
  let startBlockChunk = startblock
  do {
    endBlockChunk =
      endBlock > startBlockChunk + LOG_FILTER_CHUNK - 1 ? startBlockChunk + LOG_FILTER_CHUNK - 1 : endBlock
    const eventsChunk = await contract.queryFilter(filter, startBlockChunk, endBlockChunk)
    events.push(...eventsChunk)
    startBlockChunk = endBlockChunk + 1
  } while (endBlockChunk < endBlock)
  return events
}
