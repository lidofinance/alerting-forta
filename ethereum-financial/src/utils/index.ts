import BigNumber from "bignumber.js";

const TEN_TO_18 = new BigNumber(10).pow(18);

export function formatEth(amount: any, dp: number): string {
  return new BigNumber(String(amount)).div(TEN_TO_18).toFixed(dp);
}

export function formatDelay(fullDelaySec: number) {
  let sign = fullDelaySec >= 0 ? 1 : -1;
  let delayHours = 0;
  let delayMin = Math.floor((sign * fullDelaySec) / 60);
  let delaySec = sign * fullDelaySec - delayMin * 60;
  if (delayMin >= 60) {
    delayHours = Math.floor(delayMin / 60);
    delayMin -= delayHours * 60;
  }
  return (
    (sign == 1 ? "" : "-") +
    (delayHours > 0 ? `${delayHours} hrs ` : "") +
    (delayMin > 0 ? `${delayMin} min ` : "") +
    `${delaySec} sec`
  );
}
