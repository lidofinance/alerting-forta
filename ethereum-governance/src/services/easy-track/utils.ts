import { SAFES } from 'constants/easy-track'
import { STONKS } from 'constants/stonks'
import { getSafeAddress, etherscanAddress } from '../../shared/string'
import { Blockchain } from '../../shared/contracts'

export const getMotionCreatorNamedLink = (address: string) => {
  const ethSafes = SAFES.Ethereum
  const safeName = ethSafes.find((safeData: string[]) => safeData[0].toLowerCase() === address.toLowerCase())
  if (safeName) {
    return getSafeAddress(address, safeName[1])
  } else {
    return etherscanAddress(address)
  }
}

export const getStonksContractInfo = (address: string) => {
  return STONKS.find((c) => c.address === address)
}

export const getSafeNameByAddress = (address: string) => {
  let safeName = address
  Object.keys(SAFES).forEach((key) => {
    const safe = SAFES[key as Blockchain].find(
      (safeData: string[]) => safeData[0].toLowerCase() === address.toLowerCase(),
    )
    if (safe) {
      safeName = safe[1]
      return
    }
  })
  return safeName
}
