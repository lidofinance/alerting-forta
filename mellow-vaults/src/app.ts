import { getEthersProvider, fetchJwt, Finding, verifyJwt } from 'forta-agent'
import { LIDO_STETH_ADDRESS } from 'constants/common'
import { ETHProvider } from './clients/eth_provider'
import { FormatterWithEIP1898 } from './clients/eth_formatter'
import { DataRW } from './shared/mutex'
import * as Winston from 'winston'
import * as E from 'fp-ts/Either'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/HealthChecker.srv'
import { LidoDAO__factory } from './generated'
import { VaultWatcherSrv } from './services/vault-watcher/VaultWatcher.srv'
import { MultisigWatcherSrv } from './services/multisig-watcher/MultisigWatcher.srv'

export type Container = {
  ethClient: ETHProvider
  VaultWatcherSrv: VaultWatcherSrv
  MultisigWatcherSrv: MultisigWatcherSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getJwt(): Promise<E.Either<Error, string>> {
    let token: string
    try {
      token = await fetchJwt({})
    } catch (e) {
      return E.left(new Error(`Could not fetch jwt. Cause ${e}`))
    }

    if (process.env.NODE_ENV === 'production') {
      try {
        const isTokenOk = await verifyJwt(token)
        if (!isTokenOk) {
          return E.left(new Error(`Token verification failed`))
        }
      } catch (e) {
        return E.left(new Error(`Token verification failed`))
      }
    }

    return E.right(token)
  }

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()
      ethersProvider.formatter = new FormatterWithEIP1898()

      const stethContract = LidoDAO__factory.connect(LIDO_STETH_ADDRESS, ethersProvider)

      const ethClient = new ETHProvider(ethersProvider, stethContract)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const vaultWatcherSrv = new VaultWatcherSrv(logger, ethClient)
      const multisigWatcherSrv = new MultisigWatcherSrv(logger)

      App.instance = {
        ethClient: ethClient,
        VaultWatcherSrv: vaultWatcherSrv,
        MultisigWatcherSrv: multisigWatcherSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
