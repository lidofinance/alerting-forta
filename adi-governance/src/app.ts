import { fetchJwt, Finding, getEthersProvider, verifyJwt } from 'forta-agent'
import { ETHProvider } from './clients/eth_provider'
import { DataRW } from './utils/mutex'
import * as Winston from 'winston'
import * as E from 'fp-ts/Either'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/health-checker.srv'
import { BnbAdiSrv } from './services/bnb-adi/BnbAdi.srv'
import { CrossChainController__factory } from './generated'
import { ADI_CROSS_CHAIN_CONTROLLER } from './utils/constants'

export type Container = {
  ethClient: ETHProvider
  BnbAdiSrv: BnbAdiSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getJwt(): Promise<E.Either<Error, string>> {
    let token: string
    try {
      token = await fetchJwt({})
    } catch (e) {
      return E.left(new Error(`Could not fetch jwt. Cause ${e}`))
    }

    if (process.env.NODE_ENV === 'production') {
      try {
        const isTokenOk = await verifyJwt(token)
        if (!isTokenOk) {
          return E.left(new Error(`Token verification failed`))
        }
      } catch (e) {
        return E.left(new Error(`Token verification failed`))
      }
    }

    return E.right(token)
  }

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()

      const crossChainControllerContract = CrossChainController__factory.connect(
        ADI_CROSS_CHAIN_CONTROLLER,
        ethersProvider,
      )

      const ethClient = new ETHProvider(ethersProvider, crossChainControllerContract)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const bnbAdiSrv = new BnbAdiSrv(logger, ethClient)

      App.instance = {
        ethClient: ethClient,
        BnbAdiSrv: bnbAdiSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
