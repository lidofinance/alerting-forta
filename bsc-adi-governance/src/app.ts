import { Finding, getEthersProvider } from 'forta-agent'
import * as Winston from 'winston'
import { BSCProvider } from './clients/bsc_provider'
import { EventWatcherSrv } from './services/event-watcher/EventWatcher.srv'
import { CrossChainControllerSrv } from './services/cross-chain-controller/CrossChainController.srv'
import { AclChangesSrv } from './services/acl-changes/AclChanges.srv'
import { DataRW } from './utils/mutex'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/health-checker.srv'
import { CROSS_CHAIN_CONTROLLER_ADDRESS, CROSS_CHAIN_EXECUTOR_ADDRESS } from './utils/constants'
import { CrossChainController__factory } from './generated'
import { getCrossChainExecutorEvents } from './utils/events/cross_chain_executor_events'

export type Container = {
  bscClient: BSCProvider
  crossChainControllerSrv: CrossChainControllerSrv
  crossChainExecutorWatcherSrv: EventWatcherSrv
  bscAclChangesSrv: AclChangesSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()

      const crossChainControllerRunner = CrossChainController__factory.connect(
        CROSS_CHAIN_CONTROLLER_ADDRESS,
        ethersProvider,
      )

      const bscClient = new BSCProvider(ethersProvider, crossChainControllerRunner)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const crossChainControllerSrv = new CrossChainControllerSrv(logger, bscClient, CROSS_CHAIN_CONTROLLER_ADDRESS)
      const crossChainExecutorWatcherSrv = new EventWatcherSrv(
        'CrossChainExecutorWatcher',
        getCrossChainExecutorEvents(CROSS_CHAIN_EXECUTOR_ADDRESS),
        logger,
      )
      const bscAclChangesSrv = new AclChangesSrv(logger, bscClient)

      App.instance = {
        bscClient,
        crossChainControllerSrv,
        crossChainExecutorWatcherSrv,
        bscAclChangesSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
