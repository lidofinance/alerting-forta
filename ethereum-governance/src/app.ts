import { ethers, fetchJwt, Finding, getEthersProvider, verifyJwt } from 'forta-agent'
import {
  ARAGON_VOTING_ADDRESS,
  CURATED_NODE_OPERATORS_REGISTRY_ADDRESS,
  ENS_BASE_REGISTRAR_ADDRESS,
} from 'constants/common'
import { ETHProvider } from './clients/eth_provider'
import { FormatterWithEIP1898 } from './clients/eth_formatter'
import {
  ENS__factory,
  IncreaseStakingLimit__factory,
  NodeOperatorsRegistry__factory,
  AragonVoting__factory,
} from './generated'
import { EnsNamesSrv } from './services/ens-names/EnsNames.srv'
import { EasyTrackSrv } from './services/easy-track/EasyTrack.srv'
import { ProxyWatcherSrv } from './services/proxy-watcher/ProxyWatcher.srv'
import { DataRW } from './shared/mutex'
import * as Winston from 'winston'
import * as E from 'fp-ts/Either'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/health-checker.srv'
import { INCREASE_STAKING_LIMIT_ADDRESS } from 'constants/easy-track'
import { AclChangesSrv } from './services/acl-changes/AclChanges.srv'
import { AragonVotingSrv } from './services/aragon-voting/AragonVoting.srv'
import { TrpChangesSrv } from './services/trp-changes/TrpChanges.srv'
import { StonksSrv } from './services/stonks/Stonks.srv'
import { CrossChainWatcherSrv } from './services/cross-chain-watcher/CrossChainWatcher.srv'

export type Container = {
  ethClient: ETHProvider
  EnsNamesSrv: EnsNamesSrv
  EasyTrackSrv: EasyTrackSrv
  AclChangesSrv: AclChangesSrv
  ProxyWatcherSrv: ProxyWatcherSrv
  AragonVotingSrv: AragonVotingSrv
  TrpChangesSrv: TrpChangesSrv
  StonksSrv: StonksSrv
  CrossChainWatcherSrv: CrossChainWatcherSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getJwt(): Promise<E.Either<Error, string>> {
    let token: string
    try {
      token = await fetchJwt({})
    } catch (e) {
      return E.left(new Error(`Could not fetch jwt. Cause ${e}`))
    }

    if (process.env.NODE_ENV === 'production') {
      try {
        const isTokenOk = await verifyJwt(token)
        if (!isTokenOk) {
          return E.left(new Error(`Token verification failed`))
        }
      } catch (e) {
        return E.left(new Error(`Token verification failed`))
      }
    }

    return E.right(token)
  }

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const etherscanKey = Buffer.from('SVZCSjZUSVBXWUpZSllXSVM0SVJBSlcyNjRITkFUUjZHVQ==', 'base64').toString('utf-8')
      const ethersProvider = getEthersProvider()
      ethersProvider.formatter = new FormatterWithEIP1898()

      const etherscanProvider = new ethers.providers.EtherscanProvider(ethersProvider.network, etherscanKey)

      const ensContact = ENS__factory.connect(ENS_BASE_REGISTRAR_ADDRESS, ethersProvider)
      const increaseStakingLimitContact = IncreaseStakingLimit__factory.connect(
        INCREASE_STAKING_LIMIT_ADDRESS,
        ethersProvider,
      )
      const nodeOperatorsRegistryContract = NodeOperatorsRegistry__factory.connect(
        CURATED_NODE_OPERATORS_REGISTRY_ADDRESS,
        ethersProvider,
      )
      const aragonVotingContract = AragonVoting__factory.connect(ARAGON_VOTING_ADDRESS, ethersProvider)

      const ethClient = new ETHProvider(
        ethersProvider,
        etherscanProvider,
        ensContact,
        increaseStakingLimitContact,
        nodeOperatorsRegistryContract,
        aragonVotingContract,
      )

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const proxyWatcherSrv = new ProxyWatcherSrv(logger, ethClient)

      const ensNamesSrv = new EnsNamesSrv(logger, ethClient)
      const easyTrackSrv = new EasyTrackSrv(logger, ethClient)
      const aclChangesSrv = new AclChangesSrv(logger, ethClient)
      const aragonVotingSrv = new AragonVotingSrv(logger, ethClient)
      const trpChangesSrv = new TrpChangesSrv(logger)
      const stonksSrv = new StonksSrv(logger, ethClient)
      const crossChainWatcherSrv = new CrossChainWatcherSrv(logger, ethClient)

      App.instance = {
        ethClient: ethClient,
        EnsNamesSrv: ensNamesSrv,
        EasyTrackSrv: easyTrackSrv,
        AclChangesSrv: aclChangesSrv,
        ProxyWatcherSrv: proxyWatcherSrv,
        AragonVotingSrv: aragonVotingSrv,
        TrpChangesSrv: trpChangesSrv,
        StonksSrv: stonksSrv,
        CrossChainWatcherSrv: crossChainWatcherSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
