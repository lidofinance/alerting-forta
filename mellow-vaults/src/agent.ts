import {
  BlockEvent,
  Finding,
  FindingSeverity,
  FindingType,
  HandleBlock,
  HealthCheck,
  HandleTransaction,
} from 'forta-agent'
import * as process from 'process'
import { argv } from 'process'
import { InitializeResponse } from 'forta-agent/dist/sdk/initialize.response'
import { Initialize } from 'forta-agent/dist/sdk/handlers'
import * as E from 'fp-ts/Either'
import { App } from './app'
import { elapsedTime } from './shared/time'
import Version from './shared/version'
import { Metadata } from './entity/metadata'
import { TransactionEvent } from 'forta-agent'
import { BlockDto } from './entity/events'

export function initialize(): Initialize {
  const metadata: Metadata = {
    'version.commitHash': Version.commitHash,
    'version.commitMsg': Version.commitMsg,
  }

  return async function (): Promise<InitializeResponse | void> {
    const startTime = new Date().getTime()
    const app = await App.getInstance()

    const latestBlock = await app.ethClient.getStartedBlockForApp(argv)
    if (E.isLeft(latestBlock)) {
      console.error(`Error: ${latestBlock.left.message}`)
      console.error(`Stack: ${latestBlock.left.stack}`)

      process.exit(1)
    }

    const lastBlockNumber = latestBlock.right.number
    const [vaultWatcherSrvErr] = await Promise.all([
      app.VaultWatcherSrv.initialize(lastBlockNumber),
      app.MultisigWatcherSrv.initialize(lastBlockNumber),
      app.AclChangesSrv.initialize(lastBlockNumber),
    ])

    if (vaultWatcherSrvErr !== null) {
      console.error(`Error: ${vaultWatcherSrvErr.message}`)
      console.error(`Stack: ${vaultWatcherSrvErr.stack}`)

      process.exit(1)
    }

    const agents = [app.VaultWatcherSrv.getName(), app.MultisigWatcherSrv.getName(), app.AclChangesSrv.getName()]
    metadata.agents = '[' + agents.toString() + ']'

    await app.findingsRW.write([
      Finding.fromObject({
        name: `Agent launched`,
        description: `Version: ${Version.desc}`,
        alertId: 'LIDO-AGENT-LAUNCHED',
        severity: FindingSeverity.Info,
        type: FindingType.Info,
        metadata,
      }),
    ])

    console.log(elapsedTime('Agent.initialize', startTime) + '\n')
  }
}

export const handleBlock = (): HandleBlock => {
  return async function (blockEvent: BlockEvent): Promise<Finding[]> {
    console.log(`#ETH block: ${blockEvent.block.number}`)
    const startTime = new Date().getTime()

    const app = await App.getInstance()

    const findings: Finding[] = []
    const findingsAsync = await app.findingsRW.read()
    if (findingsAsync.length > 0) {
      findings.push(...findingsAsync)
    }

    const blockDtoEvent: BlockDto = {
      number: blockEvent?.block?.number,
      timestamp: blockEvent?.block?.timestamp,
      parentHash: blockEvent?.block?.parentHash,
    }

    const vaultWatcherSrvFindings = await app.VaultWatcherSrv.handleBlock(blockDtoEvent)
    const aclChangesSrvFindings = await app.AclChangesSrv.handleBlock(blockDtoEvent)
    findings.push(...vaultWatcherSrvFindings, ...aclChangesSrvFindings)

    console.log(elapsedTime('handleBlock', startTime) + '\n')
    return findings
  }
}

export const handleTransaction = (): HandleTransaction => {
  return async function (txEvent: TransactionEvent): Promise<Finding[]> {
    const app = await App.getInstance()

    const vaultFindings = await app.VaultWatcherSrv.handleTransaction(txEvent)
    const multisigFindings = await app.MultisigWatcherSrv.handleTransaction(txEvent)

    const findings = [...vaultFindings, ...multisigFindings]
    app.healthChecker.check(findings)

    return findings
  }
}

export const healthCheck = (): HealthCheck => {
  return async function (): Promise<string[] | void> {
    const app = await App.getInstance()

    if (!app.healthChecker.isHealth()) {
      return ['There is too much network errors']
    }

    return []
  }
}

export default {
  initialize: initialize(),
  handleBlock: handleBlock(),
  handleTransaction: handleTransaction(),
  healthCheck: healthCheck(),
}
