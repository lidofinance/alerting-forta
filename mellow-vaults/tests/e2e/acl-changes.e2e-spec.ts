import { Finding, getEthersProvider } from 'forta-agent'
import { JsonRpcProvider } from '@ethersproject/providers'

import { App } from '../../src/app'
import { etherBlockToFortaBlockEvent } from './utils'

const TEST_TIMEOUT = 60_000 // ms

describe('vault-acl-changes e2e tests', () => {
  let ethProvider: JsonRpcProvider
  // let getRoleMembers: (address: string, hash: string, currentBlock: BlockTag) => Promise<E.Either<Error, string[]>>
  let runTransaction: (txHash: string, initBlock?: number) => Promise<Finding[]>
  let runBlock: (blockHashOrNumber: string | number, initBlock?: number) => Promise<Finding[]>
  let logSpy: jest.SpyInstance
  let timeSpy: jest.SpyInstance

  beforeAll(() => {
    ethProvider = getEthersProvider()

    logSpy = jest.spyOn(console, 'log')
    logSpy.mockImplementation(() => {})
    timeSpy = jest.spyOn(Date, 'now')
    timeSpy.mockImplementation(() => new Date('2023-12-31'))

    runBlock = async (blockHashOrNumber, initBlock) => {
      const app = await App.getInstance()

      const blockNumber = initBlock
      const block = await ethProvider.getBlock(blockHashOrNumber)
      if (blockNumber) {
        await app.AclChangesSrv.initialize(blockNumber)
      }
      const blockEvent = etherBlockToFortaBlockEvent(block)
      return app.AclChangesSrv.handleBlock(blockEvent)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should everything is fine',
    async () => {
      const findings = await runBlock(20061165, 20061166)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
