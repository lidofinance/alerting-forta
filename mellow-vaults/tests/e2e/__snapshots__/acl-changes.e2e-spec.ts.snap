// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vault-acl-changes e2e tests should everything is fine 1`] = `[]`;
